src/heaps/binomial_heap.jl

```diff
@@ -0,0 +1,368 @@
+# Binomial Heap
+mutable struct MutableBinomialHeapNode{T} 
+    data::T
+    degree::Int
+    handle::Int
+    parent::Union{MutableBinomialHeapNode,Nothing}
+    child::Union{MutableBinomialHeapNode,Nothing}
+    sibling::Union{MutableBinomialHeapNode,Nothing}
+end
+MutableBinomialHeapNode(data::T, handle::Int) where T = MutableBinomialHeapNode{T}(data, 0, handle, nothing, nothing, nothing)
+#################################################
+#
+#   heap type and constructors
+#
+#################################################
+mutable struct MutableBinomialHeap{T} <: AbstractHeap{T}
+    rootList::Vector{Union{Nothing,MutableBinomialHeapNode{T}}}
+    nodecount::Int
+    cumm_nodecount::Int
+    MutableBinomialHeap{T}() where T = new{T}([], 0, 0)
+
+    function MutableBinomialHeap{T}(xs::AbstractVector{T}) where T
+        BinomialHeap = _make_binomial_heap(xs, T)
+        new{T}(BinomialHeap.rootList, BinomialHeap.nodecount, BinomialHeap.cumm_nodecount)
+    end
+    
+end
+const BinomialHeap{T} = MutableBinomialHeap{T}
+BinomialHeap(xs::AbstractVector{T}) where T = MutableBinomialHeap{T}(xs)
+
+struct HeapBoundsError <: Exception
+    msg::AbstractString
+end
+
+function show(io::IO, _heap::MutableBinomialHeap{T}) where T
+    print(io, "MutableBinomialHeap")
+    list = _heap.rootList
+    n = length(list)
+    if n == 0
+        print(io, "{$(T)}")
+    end
+    print(io, "(")
+    for i = 1:n
+        printTree(io, list[i])
+    end
+    print(io, ")")
+end
+
+function printTree(io::IO, h::Union{MutableBinomialHeapNode{T},Nothing}) where T
+    while h !== nothing 
+        print(io, h.data)
+        print(io, ",")
+        printTree(io, h.child) 
+        h = h.sibling; 
+    end
+end
+#################################################
+#
+#   core implementation
+#
+#################################################
+# make Binomial_heap from Vector
+function _make_binomial_heap(xs::AbstractVector{T}, Typ::Type{T}) where T
+    len = length(xs)
+    BinomialHeap = MutableBinomialHeap{Typ}()
+    for i = 1:len
+        push!(BinomialHeap, xs[i])
+    end
+    BinomialHeap
+end
```

Comment:
Just a tip regarding the iteration and the type specification.
```
function _make_binomial_heap(xs::AbstractVector{T}, ::Type{T}) where T
    BinomialHeap = MutableBinomialHeap{T}()
    for i in xs
        push!(BinomialHeap, i)
    end
    return BinomialHeap
end
```

Reply:
thanks, this is surely better.